### PR TITLE
[WIP] Use nep29 to help with the Python 3.6 -> 3.9 transition

### PR DESCRIPTION
**Pull request by hmaarrfk**
## Description

Migrates builds from python 3.6 to python 3.9 using nep29 to bump minimum version requirements.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by jni** [comment]
Nice, thanks @hmaarrfk! Please let me know if there is anything you'd like to discuss here. It looks to me like Py3.9 will need to be side-loaded rather than using the CI service VMs?

**Comment by grlee77** [comment]
>  It looks to me like Py3.9 will need to be side-loaded rather than using the CI service VMs?

As of a week or two ago, both Appveyor and Travis have Python 3.9 available directly. For Appveyor, you might have to select the "Visual Studio 2019" image (https://www.appveyor.com/updates/2020/10/26/)

**Comment by hmaarrfk** [comment]
It seems osx is timing out. Time to move to azure.????

**Comment by grlee77** [comment]
> It seems osx is timing out. Time to move to azure.????

Given that Travis recently announced that [OSX builds on Travis are no longer going to be free](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing), it doesn't seem that we will be able to keep them as is in any case.

There are also new restrictions to 1000 minutes on Linux builds for free accounts. I don't know what our monthly usage is now, but I don't think 1000 minutes gets us very far. I think SciPy is planning to switch existing Travis linux builds to GitHub Actions.

**Comment by hmaarrfk** [comment]
@jni I hope you aren't waiting on this PR to release the next version.

This infrastructure stuff just likes to grow in scope.

**Comment by hmaarrfk** [comment]
It seems builds are failing due to missing optional dependencies for python 3.9. i suggest skipping them for now.

I'll try to work on that later. I think we have all the environment variables in place for that.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -35,8 +26,16 @@ jobs:
...
         PIP_FLAGS: ''
         TEST_EXAMPLES: 'true'
-      Python37-x64-pre:
-        PYTHON_VERSION: '3.7'
+      Python39-x64:
+        PYTHON_VERSION: '3.9'
+        ARCH: 'x64'
+        PIP_FLAGS: ''
+        # Don't test examples since many dependencies are not available for
+        # Python 3.9
```
Comment (soupault) [inline azure-pipelines.yml:34]:
```suggestion
        # TODO: Change to `true` when all the dependencies are available for
        # Python 3.9 (as of November 2020, multiple are still missing)
```

```suggestion
        # ```

**Review by soupault — COMMENTED** [review]
Minor suggestion on changing the note. Other than that, I am +1 on merging this PR and moving to the next step.

Travis is failing on few builds. Probably related:
```
# (in `test_thresholding`)
>       pytest.importorskip('dask', reason="dask python library is not installed")
E       TypeError: importorskip() got an unexpected keyword argument 'reason'
```

**Comment by grlee77** [comment]
This has become outdated (no Travis CI anymore, already dropped Python 3.6 and added 3.9).

There are a handful of requirement .txt file increments (`seaborn`, `networkx`, etc.) that could be proposed in a standalone PR if desired.

**Pull request closed without merging**